File: src/core/url-loader.ts
export interface LoadedUrl {
  text: string
  url: string
  title?: string
}

export async function loadUrl(url: string): Promise<LoadedUrl> {
  const controller = new AbortController()
  const timeout = setTimeout(() => controller.abort(), 15000)

  try {
    const res = await fetch(url, {
      signal: controller.signal,
      headers: {
        'User-Agent': 'ContextLane/0.1.0 (context ingestion; +https://github.com/talocode/contextlane)',
        'Accept': 'text/html,application/xhtml+xml',
      },
    })
    if (!res.ok) throw new Error(`HTTP ${res.status}: ${res.statusText}`)
    const html = await res.text()
    const text = extractText(html)
    const title = extractTitle(html)
    return { text, url, title }
  } finally {
    clearTimeout(timeout)
  }
}

function extractText(html: string): string {
  let text = html

  // Remove unwanted elements
  text = text.replace(/<script[^>]*>[\s\S]*?<\/script>/gi, ' ')
  text = text.replace(/<style[^>]*>[\s\S]*?<\/style>/gi, ' ')
  text = text.replace(/<nav[^>]*>[\s\S]*?<\/nav>/gi, ' ')
  text = text.replace(/<footer[^>]*>[\s\S]*?<\/footer>/gi, ' ')
  text = text.replace(/<header[^>]*>[\s\S]*?<\/header>/gi, ' ')
  text = text.replace(/<aside[^>]*>[\s\S]*?<\/aside>/gi, ' ')
  text = text.replace(/<form[^>]*>[\s\S]*?<\/form>/gi, ' ')
  text = text.replace(/<svg[^>]*>[\s\S]*?<\/svg>/gi, ' ')

  // Replace block elements with newlines
  text = text.replace(/<\/(p|div|h[1-6]|li|blockquote|section|article|pre|tr|th|td|br|hr|dl|dt|dd|ol|ul|table)[^>]*>/gi, '\n')

  // Remove all remaining tags
  text = text.replace(/<[^>]+>/g, ' ')

  // Decode HTML entities
  text = text.replace(/&nbsp;/g, ' ')
    .replace(/&amp;/g, '&')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/&#x27;/g, "'")
    .replace(/&#x2F;/g, '/')
    .replace(/&#\d+;/g, ' ')

  // Clean up whitespace
  text = text
    .replace(/\r\n/g, '\n')
    .replace(/\r/g, '\n')
    .replace(/\n{3,}/g, '\n\n')
    .replace(/[ \t]+/g, ' ')
    .replace(/^\s+|\s+$/gm, '')
    .trim()

  // Remove very short lines (likely noise)
  const lines = text.split('\n').filter(l => l.trim().length > 2)
  text = lines.join('\n')

  return text
}

function extractTitle(html: string): string | undefined {
  const m = html.match(/<title[^>]*>([^<]+)<\/title>/i)
  if (m) return m[1].trim().replace(/\s+/g, ' ')

  const og = html.match(/<meta[^>]+property="og:title"[^>]+content="([^"]+)"/i)
  if (og) return og[1].trim()

  const h1 = html.match(/<h1[^>]*>([^<]+)<\/h1>/i)
  if (h1) return h1[1].trim()

  return undefined
}

File: python/contextlane/core/url_loader.py
import re
import urllib.request
import urllib.error


def load_url(url: str) -> dict:
    req = urllib.request.Request(
        url,
        headers={
            "User-Agent": "ContextLane/0.2.0 (context ingestion; +https://github.com/talocode/contextlane)",
            "Accept": "text/html,application/xhtml+xml",
        },
    )
    try:
        with urllib.request.urlopen(req, timeout=15) as resp:
            html = resp.read().decode("utf-8", errors="replace")
    except urllib.error.HTTPError as e:
        raise RuntimeError(f"HTTP {e.code}: {e.reason}")
    except urllib.error.URLError as e:
        raise RuntimeError(f"URL error: {e.reason}")

    text = _extract_text(html)
    title = _extract_title(html)
    return {"text": text, "url": url, "title": title}


def _extract_text(html: str) -> str:
    text = html
    for tag in ["script", "style", "nav", "footer", "header", "aside", "form", "svg"]:
        text = re.sub(f"<{tag}[^>]*>.*?</{tag}>", " ", text, flags=re.DOTALL | re.IGNORECASE)

    text = re.sub(r"<\/(p|div|h[1-6]|li|blockquote|section|article|pre|tr|th|td|br|hr)\s*>", "\n", text, flags=re.IGNORECASE)
    text = re.sub(r"<[^>]+>", " ", text)

    # entities
    text = text.replace("&nbsp;", " ").replace("&amp;", "&").replace("&lt;", "<").replace("&gt;", ">")
    text = text.replace("&quot;", '"').replace("&#39;", "'").replace("&#x27;", "'").replace("&#x2F;", "/")
    text = re.sub(r"&#\d+;", " ", text)

    text = re.sub(r"\r\n", "\n", text)
    text = re.sub(r"\r", "\n", text)
    text = re.sub(r"\n{3,}", "\n\n", text)
    text = re.sub(r"[ \t]+", " ", text)

    lines = [l.strip() for l in text.split("\n") if len(l.strip()) > 2]
    return "\n".join(lines)


def _extract_title(html: str) -> str | None:
    m = re.search(r"<title[^>]*>([^<]+)</title>", html, re.IGNORECASE)
    if m:
        return m.group(1).strip()
    m = re.search(r'<meta[^>]+property="og:title"[^>]+content="([^"]+)"', html, re.IGNORECASE)
    if m:
        return m.group(1).strip()
    m = re.search(r"<h1[^>]*>([^<]+)</h1>", html, re.IGNORECASE)
    if m:
        return m.group(1).strip()
    return None
